algorithms/src/std_algorithms/Kokkos_Constraints.hpp

```diff
@@ -0,0 +1,225 @@
+struct are_random_access_iterators;
+
+template <class T>
+struct are_random_access_iterators<T> {
+  static constexpr bool value =
+      is_iterator<T>::value &&
+      std::is_same<typename T::iterator_category,
+                   std::random_access_iterator_tag>::value;
+};
+
+template <class Head, class... Tail>
+struct are_random_access_iterators<Head, Tail...> {
+  static constexpr bool value = are_random_access_iterators<Head>::value &&
+                                are_random_access_iterators<Tail...>::value;
+};
+
+//
+// are_accessible_iterators
+//
+template <class... Args>
+struct are_accessible_iterators;
+
+template <class ExeSpace, class IteratorType>
+struct are_accessible_iterators<ExeSpace, IteratorType> {
+  using view_type = typename IteratorType::view_type;
+  static constexpr bool value =
+      SpaceAccessibility<ExeSpace,
+                         typename view_type::memory_space>::accessible;
+};
+
+template <class ExeSpace, class Head, class... Tail>
+struct are_accessible_iterators<ExeSpace, Head, Tail...> {
+  static constexpr bool value =
+      are_accessible_iterators<ExeSpace, Head>::value &&
+      are_accessible_iterators<ExeSpace, Tail...>::value;
+};
+
+template <class ExecutionSpace, class... IteratorTypes>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_random_access_and_accessible() {
+  static_assert(
+      are_random_access_iterators<IteratorTypes...>::value,
+      "Currently, Kokkos standard algorithms require random access iterators.");
+
+  static_assert(
+      are_accessible_iterators<ExecutionSpace, IteratorTypes...>::value,
+      "Incompatible view/iterator and execution space");
+}
+
+//
+// have matching difference_type
+//
+template <class... Args>
+struct iterators_have_matching_difference_type;
+
+template <class T>
+struct iterators_have_matching_difference_type<T> {
+  static constexpr bool value = true;
+};
+
+template <class T1, class T2>
+struct iterators_have_matching_difference_type<T1, T2> {
+  static constexpr bool value =
+      std::is_same<typename T1::difference_type,
+                   typename T2::difference_type>::value;
+};
+
+template <class T1, class T2, class... Tail>
+struct iterators_have_matching_difference_type<T1, T2, Tail...> {
+  static constexpr bool value =
+      iterators_have_matching_difference_type<T1, T2>::value &&
+      iterators_have_matching_difference_type<T2, Tail...>::value;
+};
+
+template <class... IteratorTypes>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_iterators_have_matching_difference_type() {
+  static_assert(
+      iterators_have_matching_difference_type<IteratorTypes...>::value,
+      "Iterators do not have matching difference_type");
+}
+
+//
+// not_openmptarget
+//
+template <class ExeSpace>
+struct not_openmptarget {
+#if not defined KOKKOS_ENABLE_OPENMPTARGET
+  static constexpr bool value = true;
+#else
+  static constexpr bool value =
+      !std::is_same<std::remove_cv_t<std::remove_reference_t<ExeSpace> >,
```

Comment:
Note that we have `Kokkos::Impl::remove_cvref_t`.  Also I am not convinced you should be removing the reference.

Reply:
My understanding would be to use `std::decay_t` here to get the basic type.

Reply:
Also, shouldn't this be in an `Impl` namespace?